pkg/server/server.go

```diff
@@ -53,51 +39,61 @@ type Config struct {
 type Server struct {
 	Catalog    catalog.Catalog
 	Config     *Config
-	endpoints  endpoint.Endpoint
+	endpoints  endpoints.Server
 	privateKey *ecdsa.PrivateKey
 	svid       *x509.Certificate
+
+	m *sync.RWMutex
+	t *tomb.Tomb
 }
 
 // Run the server
 // This method initializes the server, including its plugins,
-// and then blocks on the main event loop.
+// and then blocks until it's shut down or an error is encountered.
 func (server *Server) Run() error {
+	if server.t == nil {
+		server.t = new(tomb.Tomb)
+	}
+
+	server.t.Go(server.run)
+
+	return server.t.Wait()
+}
+
+func (server *Server) run() error {
 	server.prepareUmask()
 
+	if server.m == nil {
+		server.m = new(sync.RWMutex)
```

Comment:
Just out of curiosity, why is this done here?

Reply:
Ideally we would have a `New` function, but I stopped short of adding one cause I felt this PR already had enough changes as is :)

I have moved this line to `Run` function for consistency.